### PR TITLE
Allow notEmpty and allowEmpty to get fields as list

### PR DESCRIPTION
**Pull request by Graziel**
see #8659

### CODE REVIEW

File: src/Validation/Validator.php [hunk 1]
```diff
@@ -526,6 +582,19 @@ public function allowEmpty($field, $when = true)
      * $validator->notEmpty('email'); // Email cannot be empty
      * $validator->notEmpty('email', $message, 'create'); // Email can be empty on update
      * $validator->notEmpty('email', $message, 'update'); // Email can be empty on create
+     * $validator->notEmpty(['email', 'title'], $message, 'update');
+     * // Email and title can be empty on create but can not on update
+     * $validator->notEmpty(
+     *      [
+     *          'email',
+     *          'title' => [
+     *              'when' => true,
```
Comment (markstory) [inline src/Validation/Validator.php]: In reading this, its not clear what `when => true` will do.

File: src/Validation/Validator.php [hunk 2]
```diff
@@ -619,9 +711,10 @@ public function lengthBetween($field, array $range, $message = null, $when = nul
             throw new InvalidArgumentException('The $range argument requires 2 numbers');
         }
         $extra = array_filter(['on' => $when, 'message' => $message]);
+
         return $this->add($field, 'lengthBetween', $extra + [
-            'rule' => ['lengthBetween', array_shift($range), array_shift($range)],
-        ]);
+                'rule' => ['lengthBetween', array_shift($range), array_shift($range)],
+            ]);
```
Comment (markstory) [inline src/Validation/Validator.php]: Why are all these methods changing?

File: src/Validation/Validator.php (hunk 1 above)
Comment (markstory) [inline src/Validation/Validator.php]: In reading the rest of this, its the 'on' condition. Any reason to not leave it named 'on'?

Comment (Graziel) [inline src/Validation/Validator.php]: this is how its been in code (https://github.com/cakephp/cakephp/blob/3.next/src/Validation/Validator.php#L509) i also dont know why its not 'on' but 'when' so ive left just left it same, i can change it for consistency if you want

File: src/Validation/Validator.php (hunk 2 above)
Comment (Graziel) [inline src/Validation/Validator.php]: im using autoformating of PHPStorm and to be frank i dont know why its like that but also PHPCS didnt complain about it

File: src/Validation/Validator.php [hunk 3]
```diff
@@ -500,20 +514,62 @@ public function requirePresence($field, $mode = true, $message = null)
...
+     *
+     * @param string $fieldName name of field
+     * @param array $settingKeys keys that will be used to create default settings
+     * @param array $settingsValues values that will be used to create default settings
+     * @param array $setting settings from data
+     * @return array
+     */
+    protected function _convertValidatorToArray($fieldName, $settingKeys = [], $settingsValues = [], $setting = [])
+    {
+        if (is_string($setting)) {
```
Comment (dereuromark) [inline src/Validation/Validator.php]: string is not mentioned in doc block. Also, variable might better be $settings maybe?

File: src/Validation/Validator.php (hunk 2 above)
Comment (markstory) [inline src/Validation/Validator.php]: Can you revert the autoformatting? It makes this change much larger/noisier than it needs to be.

File: src/Validation/Validator.php (hunk 1 above)
Comment (markstory) [inline src/Validation/Validator.php]: No lets leave it as it. Its just my brain being wrong.

File: src/Validation/Validator.php (hunk 3 above)
Comment (Graziel) [inline src/Validation/Validator.php]: good catch

**Comment by markstory** [comment]
@Graziel You'll need to solve the merge conflicts before this can be merged.

**Comment by Graziel** [comment]
bump for glory

File: src/Validation/Validator.php [hunk 4]
```diff
@@ -422,56 +422,56 @@ public function remove($field, $rule = null)
...
      * @param bool|string|callable $mode Valid values are true, false, 'create', 'update'.
      * If a callable is passed then the field will be required only when the callback
      * returns true.
      * @param string|null $message The message to show if the field presence validation fails.
+     *
      * @return $this
      */
     public function requirePresence($field, $mode = true, $message = null)
     {
+        $settingsDefault = [
```
Comment (dereuromark) [inline src/Validation/Validator.php]: I would go with $defaultSettings or just $defaults.

**Comment by markstory** [comment]
Nice work @Graziel!

**Pull request merged**